File: backend/schemas/pool.py
"""
Pool schemas - Pydantic models for fund pool management API.
"""
from pydantic import BaseModel, Field
from datetime import date, datetime
from typing import Optional, List


class PoolAddRequest(BaseModel):
    """Request schema for adding a single fund to pool."""
    pool_type: str = Field(..., description="Pool type: entry, focus, exit")
    fund_code: str = Field(..., description="Fund code")
    fund_name: str = Field(..., description="Fund name")
    notes: Optional[str] = Field(None, description="Optional notes")


class PoolBulkAddRequest(BaseModel):
    """Request schema for bulk adding funds to pool."""
    pool_type: str = Field(..., description="Pool type: entry, focus, exit")
    funds: List[PoolAddRequest] = Field(..., max_length=50, description="List of funds to add (max 50)")


class PoolTransferRequest(BaseModel):
    """Request schema for transferring fund between pools."""
    id: int = Field(..., description="Pool entry ID")
    new_pool_type: str = Field(..., description="New pool type: entry, focus, exit")


class PoolStatusUpdateRequest(BaseModel):
    """Request schema for updating pool entry status."""
    id: int = Field(..., description="Pool entry ID")
    status: str = Field(..., description="Status: active, removed")
    removed_date: Optional[date] = Field(None, description="Removed date (required if status is removed)")


class PoolFundResponse(BaseModel):
    """Response schema for a single pool fund entry."""
    id: int
    pool_type: str
    fund_code: str
    fund_name: str
    status: str
    added_date: date
    removed_date: Optional[date]
    notes: Optional[str]
    created_at: datetime
    
    class Config:
        from_attributes = True


class PoolListResponse(BaseModel):
    """Response schema for pool list."""
    pool_type: str
    funds: List[PoolFundResponse]
    total: int

File: frontend/src/api/pool.ts
/**
 * Pool API endpoints for fund pool management
 */
import api from './index'

// Types
export interface PoolFund {
  id: number
  pool_type: 'entry' | 'focus' | 'exit'
  fund_code: string
  fund_name: string
  status: 'active' | 'removed'
  added_date: string
  removed_date?: string
  notes?: string
  // Extended fields for display
  return_1y?: number
  scale?: number
  manager?: string
}

export interface PoolAddRequest {
  pool_type: string
  fund_code: string
  fund_name: string
  notes?: string
}

export interface PoolListResponse {
  funds: PoolFund[]
  total: number
}

export interface PoolBulkAddRequest {
  pool_type: string
  funds: PoolAddRequest[]
}

export interface PoolTransferRequest {
  id: number
  new_pool_type: string
}

export interface PoolStatusUpdateRequest {
  id: number
  status: string
  removed_date?: string
}

// List funds in pool (optionally filtered by pool type and status)
export const listPoolFunds = async (
  poolType?: string,
  status?: string
): Promise<PoolListResponse> => {
  const params: Record<string, string> = {}
  if (poolType) params.pool_type = poolType
  if (status) params.status = status
  const response = await api.get('/pool', { params })
  return response.data!
}

// Add a fund to pool
export const addFundToPool = async (data: PoolAddRequest): Promise<PoolFund> => {
  const response = await api.post('/pool/add', data)
  return response.data!
}

// Bulk add funds to pool
export const bulkAddFundsToPool = async (data: PoolBulkAddRequest): Promise<PoolListResponse> => {
  const response = await api.post('/pool/bulk-add', data)
  return response.data!
}

// Remove a fund from pool
export const removeFundFromPool = async (id: number): Promise<void> => {
  const response = await api.delete(`/pool/${id}`)
  return response.data!
}

// Transfer fund between pools
export const transferFundBetweenPools = async (data: PoolTransferRequest): Promise<PoolFund> => {
  const response = await api.post('/pool/transfer', data)
  return response.data!
}

// Update fund status in pool
export const updatePoolFundStatus = async (data: PoolStatusUpdateRequest): Promise<PoolFund> => {
  const response = await api.put('/pool/status', data)
  return response.data!
}

export const poolApi = {
  list: listPoolFunds,
  add: addFundToPool,
  bulkAdd: bulkAddFundsToPool,
  remove: removeFundFromPool,
  transfer: transferFundBetweenPools,
  updateStatus: updatePoolFundStatus,
}
